frontend/src/test/java/com/google/tripmeout/frontend/storage/PersistentPlaceVisitStorageTest.java

```diff
@@ -0,0 +1,165 @@
+    storage.addPlaceVisit(place1);
+    Assert.assertThrows(
+        PlaceVisitAlreadyExistsException.class, () -> storage.addPlaceVisit(place1));
+  }
+
+  @Test
+  public void getTripPlaceVisits_returnPlacesFromOneTrip() throws Exception {
+    PlaceVisitModel place1 = createPlaceWithTrip1();
+    PlaceVisitModel place2 = createPlaceWithTrip1();
+    PlaceVisitModel place3 = createPlaceWithTrip1();
+    storage.addPlaceVisit(place1);
+    storage.addPlaceVisit(place2);
+    storage.addPlaceVisit(place3);
+    assertThat(storage.getTripPlaceVisits("trip1")).containsExactly(place1, place2, place3);
+  }
+  @Test
+  public void updateUserMark_returnTripWithNewUserMark() throws Exception {
+    PlaceVisitModel place1 = createPlaceWithTrip1();
+    storage.addPlaceVisit(place1);
+    storage.updateUserMarkOrAddPlaceVisit(place1, PlaceVisitModel.UserMark.NO);
+    assertEquals(PlaceVisitModel.UserMark.NO,
+        storage.getPlaceVisit(place1.tripId(), place1.id()).get().userMark());
+  }
+
+  @Test
+  public void removePlaceVisit_returnListofTrips() throws Exception {
+    PlaceVisitModel place1 = createPlaceWithTrip1();
+    PlaceVisitModel place2 = createPlaceWithTrip1();
+    PlaceVisitModel place3 = createPlaceWithTrip1();
+    PlaceVisitModel place4 = createPlaceWithTrip1();
+    storage.addPlaceVisit(place1);
+    storage.addPlaceVisit(place2);
+    storage.addPlaceVisit(place3);
+    storage.addPlaceVisit(place4);
+    storage.removePlaceVisit(place4.tripId(), place4.id());
+    assertThat(storage.getTripPlaceVisits("trip1")).containsExactly(place1, place2, place3);
+  }
+
+  @Test
+  public void removePlaceVisit_returnPlaceNotFoundException() throws Exception {
+    PlaceVisitModel place1 = createPlaceWithTrip1();
+    PlaceVisitModel place2 = createPlaceWithTrip1();
+    storage.addPlaceVisit(place1);
+    storage.addPlaceVisit(place2);
+    String id = KeyFactory.keyToString(KeyFactory.createKey("trip2", UUID.randomUUID().toString()));
+    Assert.assertThrows(
+        PlaceVisitNotFoundException.class, () -> storage.removePlaceVisit(place1.tripId(), id));
+  }
+
+  @Test
+  public void getTripPlaceVisits_returnListofTrip1() throws Exception {
+    PlaceVisitModel place1 = createPlaceWithTrip1();
+    PlaceVisitModel place2 = createPlaceWithTrip2();
+    PlaceVisitModel place3 = createPlaceWithTrip1();
+    PlaceVisitModel place4 = createPlaceWithTrip2();
+    storage.addPlaceVisit(place1);
+    storage.addPlaceVisit(place2);
+    storage.addPlaceVisit(place3);
+    storage.addPlaceVisit(place4);
+    assertThat(storage.getTripPlaceVisits("trip1")).containsExactly(place1, place3);
+    assertThat(storage.getTripPlaceVisits("trip2")).containsExactly(place2, place4);
+  }
+
+  @Test
+  public void removeTripPlaceVisits_returnListofTrip1() throws Exception {
+    PlaceVisitModel place1 = createPlaceWithTrip1();
+    PlaceVisitModel place2 = createPlaceWithTrip2();
+    PlaceVisitModel place3 = createPlaceWithTrip1();
```

Comment:
I'm confused. Do you mean to repeat the same trips? Looks like you're doing this in a few places where the tests are not specifically expecting exceptions.

Reply:
Yes i do mean to reuse the same tripid this way I can check if the places can be sorts by trip but every place is using this method would be different